daft/execution/execution_step.py

```diff
@@ -290,6 +290,47 @@ def num_outputs(self) -> int:
         return 1
 
 
+@dataclass(frozen=True)
+class ScanWithTask(SingleOutputInstruction):
+    scan_task: ScanTask
+
+    def run(self, inputs: list[MicroPartition]) -> list[MicroPartition]:
+        return self._scan(inputs)
+
+    def _scan(self, inputs: list[MicroPartition]) -> list[MicroPartition]:
+        assert len(inputs) == 0
+        table = MicroPartition._from_scan_task(self.scan_task)
+        return [table]
+
+    def run_partial_metadata(self, input_metadatas: list[PartialPartitionMetadata]) -> list[PartialPartitionMetadata]:
+        assert len(input_metadatas) == 0
+
+        return [
+            PartialPartitionMetadata(
+                num_rows=self.scan_task.num_rows(),
+                size_bytes=self.scan_task.size_bytes(),
+            )
+        ]
+
+
+@dataclass(frozen=True)
+class EmptyScan(SingleOutputInstruction):
+    schema: Schema
+
+    def run(self, inputs: list[MicroPartition]) -> list[MicroPartition]:
+        return [MicroPartition.empty(self.schema)]
+
+    def run_partial_metadata(self, input_metadatas: list[PartialPartitionMetadata]) -> list[PartialPartitionMetadata]:
+        assert len(input_metadatas) == 0
+
+        return [
+            PartialPartitionMetadata(
+                num_rows=0,
+                size_bytes=0,
+            )
+        ]
+
+
```

Comment:
Decided to move these to `execution_step.py` since it's a more sensible location for them, and to not introduce an otherwise unnecessary `ray_runner.py` -> `rust_physical_plan_shim.py` dependency.